server/storage/schema/actions.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"go.etcd.io/etcd/server/v3/storage/backend"
+	"go.uber.org/zap"
+)
+
+type action interface {
+	unsafeDo(tx backend.BatchTx) (revert action, err error)
+}
+
+type setKeyAction struct {
+	Bucket     backend.Bucket
+	FieldName  []byte
+	FieldValue []byte
+}
+
+func (a setKeyAction) unsafeDo(tx backend.BatchTx) (action, error) {
+	revert := currentStateFieldAction(tx, a.Bucket, a.FieldName)
+	tx.UnsafePut(a.Bucket, a.FieldName, a.FieldValue)
+	return revert, nil
+}
+
+type deleteKeyAction struct {
+	Bucket    backend.Bucket
+	FieldName []byte
+}
+
+func (a deleteKeyAction) unsafeDo(tx backend.BatchTx) (action, error) {
+	revert := currentStateFieldAction(tx, a.Bucket, a.FieldName)
+	tx.UnsafeDelete(a.Bucket, a.FieldName)
+	return revert, nil
+}
+
+func currentStateFieldAction(tx backend.BatchTx, bucket backend.Bucket, fieldName []byte) action {
```

Comment:
Please add comment

Reply:
(make it clear that it produces a rollback action)

Reply:
I think it's clearer to have in separate function, just the original name was bad. I changed the name to `restoreFieldValueAction`, please let me know what you think.